src/pybind/mgr/cephadm/module.py

```diff
@@ -1575,14 +1575,21 @@ def _daemon_action(self, daemon_type: str, daemon_id: str, host: str, action: st
 
         self._daemon_action_set_image(action, image, daemon_type, daemon_id)
 
-        if action == 'redeploy':
-            if self.daemon_is_self(daemon_type, daemon_id):
-                self.mgr_service.fail_over()
-                return ''  # unreachable
-            # stop, recreate the container+unit, then restart
-            return CephadmServe(self)._create_daemon(daemon_spec)
-        elif action == 'reconfig':
-            return CephadmServe(self)._create_daemon(daemon_spec, reconfig=True)
+        if action == 'redeploy' and self.daemon_is_self(daemon_type, daemon_id):
+            self.mgr_service.fail_over()
+            return ''  # unreachable
+
+        if action == 'redeploy' or action == 'reconfig':
+            if daemon_type != 'osd':
+                try:
+                    daemon_spec = self.cephadm_services[daemon_type_to_service(
+                        daemon_type)].prepare_create(daemon_spec)
+                except (RuntimeError, OrchestratorError) as e:
+                    msg = "Failed to {} daemon {} on {}: {}".format(
+                        action, daemon_spec.name(), daemon_spec.host, e)
+                    self.log.info(msg)
+                    return msg
```

Comment:
can you move this into a different commit and explain why this is required?

Reply:
@sebastian-philipp  I've put this in another commit and added some explanation to that commit message. Do you want some inline code comments as well?